codec/src/main/java/io/netty/handler/codec/ByteString.java

```diff
@@ -0,0 +1,756 @@
+        int max = Integer.MIN_VALUE / radix;
+        int result = 0;
+        int offset = start;
+        while (offset < end) {
+            int digit = Character.digit((char) (value[offset++] & 0xFF), radix);
+            if (digit == -1) {
+                throw new NumberFormatException(subSequence(start, end).toString());
+            }
+            if (max > result) {
+                throw new NumberFormatException(subSequence(start, end).toString());
+            }
+            int next = result * radix - digit;
+            if (next > result) {
+                throw new NumberFormatException(subSequence(start, end).toString());
+            }
+            result = next;
+        }
+        if (!negative) {
+            result = -result;
+            if (result < 0) {
+                throw new NumberFormatException(subSequence(start, end).toString());
+            }
+        }
+        return result;
+    }
+
+    public final long parseLong() {
+        return parseLong(0, length(), 10);
+    }
+
+    public final long parseLong(int radix) {
+        return parseLong(0, length(), radix);
+    }
+
+    public final long parseLong(int start, int end) {
+        return parseLong(start, end, 10);
+    }
+
+    public final long parseLong(int start, int end, int radix) {
+        if (radix < Character.MIN_RADIX || radix > Character.MAX_RADIX) {
+            throw new NumberFormatException();
+        }
+
+        if (start == end) {
+            throw new NumberFormatException();
+        }
+
+        int i = start;
+        boolean negative = byteAt(i) == '-';
+        if (negative && ++i == end) {
+            throw new NumberFormatException(subSequence(start, end).toString());
+        }
+
+        return parseLong(i, end, radix, negative);
+    }
+
+    private long parseLong(int start, int end, int radix, boolean negative) {
+        final byte[] value = this.value;
+        long max = Long.MIN_VALUE / radix;
+        long result = 0;
+        int offset = start;
+        while (offset < end) {
+            int digit = Character.digit((char) (value[offset++] & 0xFF), radix);
+            if (digit == -1) {
+                throw new NumberFormatException(subSequence(start, end).toString());
+            }
+            if (max > result) {
+                throw new NumberFormatException(subSequence(start, end).toString());
+            }
+            long next = result * radix - digit;
+            if (next > result) {
+                throw new NumberFormatException(subSequence(start, end).toString());
+            }
+            result = next;
+        }
+        if (!negative) {
+            result = -result;
+            if (result < 0) {
+                throw new NumberFormatException(subSequence(start, end).toString());
+            }
+        }
+        return result;
+    }
+
+    public final char parseChar() {
+        return parseChar(0, length());
+    }
+
+    public final char parseChar(int start, int end) {
```

Comment:
These methods were copied from `AsciiString` directly and are incorrect in this context.  I will fix this for `ByteString` and override for `AsciiString`.